dll/cpl/input/settings_page.c

```diff
@@ -11,192 +11,341 @@
+        item.pszText        = pInputNode->pLocale->pszName;
+        item.iImage         = LangImageIndex;
+        item.iSelectedImage = LangImageIndex;
+        item.lParam         = LOWORD(pInputNode->pLocale->dwId);
+        if (bBold)
+        {
+            item.state = item.stateMask = TVIS_BOLD;
+        }
+        insert.hParent      = TVI_ROOT;
+        insert.hInsertAfter = TVI_LAST;
+        insert.item         = item;
+        hItem = TreeView_InsertItem(hwndList, &insert);
+
+        // Keyboard
+        ZeroMemory(&item, sizeof(item));
+        item.mask           = TVIF_TEXT | TVIF_IMAGE | TVIF_PARAM | TVIF_SELECTEDIMAGE | TVIF_STATE;
+        item.pszText        = szKeyboard;
+        item.iImage         = KeyboardImageIndex;
+        item.iSelectedImage = KeyboardImageIndex;
+        item.lParam         = 0xFFFF;
+        if (bBold)
+        {
+            item.state = item.stateMask = TVIS_BOLD;
+        }
+        insert.hParent      = hItem;
+        insert.hInsertAfter = TVI_LAST;
+        insert.item         = item;
+        hItem = TreeView_InsertItem(hwndList, &insert);
+    }
+    else
     {
-        HICON hLayoutIcon;
+        // Language
+        ZeroMemory(&item, sizeof(item));
+        item.mask           = TVIF_STATE | TVIF_HANDLE;
+        item.hItem          = hItem;
+        item.stateMask      = TVIS_BOLD;
+        if (TreeView_GetItem(hwndList, &item) && bBold && !(item.state & TVIS_BOLD))
+        {
+            item.mask = TVIF_STATE | TVIF_HANDLE;
+            item.hItem = hItem;
+            item.state = item.stateMask = TVIS_BOLD;
+            TreeView_SetItem(hwndList, &item);
+        }
 
-        hLayoutIcon = CreateLayoutIcon(pInputNode->pszIndicator,
-                                       (pInputNode->wFlags & INPUT_LIST_NODE_FLAG_DEFAULT));
-        if (hLayoutIcon != NULL)
+        // Keyboard
+        hItem = TreeView_GetChild(hwndList, hItem);
+        ZeroMemory(&item, sizeof(item));
+        item.mask           = TVIF_STATE | TVIF_HANDLE;
+        item.hItem          = hItem;
+        item.stateMask      = TVIS_BOLD;
+        if (TreeView_GetItem(hwndList, &item) && bBold && !(item.state & TVIS_BOLD))
         {
-            ImageIndex = ImageList_AddIcon(hImageList, hLayoutIcon);
-            DestroyIcon(hLayoutIcon);
+            item.mask = TVIF_STATE | TVIF_HANDLE;
+            item.hItem = hItem;
+            item.state = item.stateMask = TVIS_BOLD;
+            TreeView_SetItem(hwndList, &item);
         }
     }
 
-    ZeroMemory(&item, sizeof(item));
-    item.mask    = LVIF_TEXT | LVIF_PARAM | LVIF_IMAGE;
-    item.pszText = pInputNode->pLocale->pszName;
-    item.iItem   = ListView_GetItemCount(hwndList);
-    item.lParam  = (LPARAM)pInputNode;
-    item.iImage  = ImageIndex;
-    ItemIndex = ListView_InsertItem(hwndList, &item);
+    // Input method
+    if (hItem)
+    {
+        // FIXME: IME icon
```

Comment:
In which aspect this needs to be fixed?

Reply:
Loading 1st resource icon from IME file...

Reply:
ok this needs to be spelled out in the comment (unless you expect to work on that in the near future)

Reply:
Added comment.